metals/src/main/scala/scala/meta/internal/metals/ScalafixProvider.scala

```diff
@@ -64,64 +68,92 @@ case class ScalafixProvider(
-          case Success(results) if !scalafixSucceded(results) =>
-            val scalafixError = getMessageErrorFromScalafix(results)
-            val exception = ScalafixRunException(scalafixError)
-            reportScalafixError(
-              scalafixError,
-              exception
-            )
+
+    compilations.compilationFinished(file).flatMap { _ =>
+      val scalafixEvaluation =
+        scalafixEvaluate(
+          file,
+          scalaTarget,
+          inBuffers.value,
+          retried || isUnsaved(inBuffers.text, fromDisk.text)
+        )
+
+      scalafixEvaluation match {
+        case Failure(exception) =>
+          reportScalafixError(
+            "Unable to run scalafix, please check logs for more info.",
+            exception
+          )
+          Future.failed(exception)
+        case Success(results)
+            if !scalafixSucceded(results) && hasStaleSemanticdb(
+              results
+            ) && buildClient.buildHasErrors(file) =>
+          val msg = "Attempt to organize your imports failed. " +
+            "It looks like you have compilation issues causing your semanticdb to be stale. " +
+            "Ensure everything is compiling and try again."
+          scribe.warn(
+            msg
+          )
+          languageClient.showMessage(
+            MessageType.Warning,
+            msg
+          )
+          Future.successful(Nil)
+        case Success(results) if !scalafixSucceded(results) =>
+          val scalafixError = getMessageErrorFromScalafix(results)
+          val exception = ScalafixRunException(scalafixError)
+          reportScalafixError(
+            scalafixError,
+            exception
+          )
+          if (!retried && hasStaleSemanticdb(results)) {
+            // Retry, since the semanticdb might be stale
+            organizeImports(file, scalaTarget, retried = true)
+          } else {
             Future.failed(exception)
-          case Success(results) =>
-            Future.successful {
-              val edits = for {
-                fileEvaluation <- results.getFileEvaluations().headOption
-                patches <- fileEvaluation.previewPatches().asScala
-              } yield textEditsFrom(patches, fromDisk)
-              edits.getOrElse(Nil)
-            }
-        }
+          }
+        case Success(results) =>
+          Future.successful {
+            val edits = for {
+              fileEvaluation <- results.getFileEvaluations().headOption
+              patches <- fileEvaluation.previewPatches().asScala
+            } yield textEditsFrom(patches, inBuffers)
+            edits.getOrElse(Nil)
+          }
+
       }
     }
   }
 
+  private def createTemporarySemanticdb(
+      file: AbsolutePath,
+      contents: String
+  ) = {
+    interactive
+      .textDocument(file, Some(contents))
+      .documentIncludingStale
+      .flatMap { semanticdb =>
+        if (semanticdb.diagnostics.exists(_.severity.isError)) {
+          None
+        } else {
```

Comment:
Here we can cleanup diagnostic in case of error before saving it to a file.
That would allow to perform imports reordering in this case.
Smth like:
```scala
val toSave = 
  if (semanticdb.diagnostics.exists(_.severity.isError))
    semanticdb.copy(diagnostic = Seq.empty)
  else
    semanticdb
```

Reply:
I don't think we want to remove the diagnostics (also it does work with them), because if the file doesn't compile the results of OrganizeImports might be unexpected. Essentially, what would happen here is that `ClassDoNotExist` will be removed while it might have just been a typo.

Reply:
The result is expected - without diagnostic it reorders imports only.
I've checked the suggested change locally and it works well.

Sometimes, when I add import manually and file has compile errors, I would like to perform organize imports right after that to put in in correct place.